### ISSUE TITLE
Setting Session to blocking in Drop for SftpInnerDropWrapper is incompatible with async usage

### ISSUE THREAD

**Issue by emesterhazy** [alexcrichton/ssh2-rs#226]
https://github.com/alexcrichton/ssh2-rs/blob/40d7d18415324205a9ba7ddd1c08b9f0fade42ec/src/sftp.rs#L503-L517

In `Drop` for `SftpInnerDropWrapper` we set the session to blocking before completing the drop. Presumably this is related to https://github.com/alexcrichton/ssh2-rs/issues/220 since in non-blocking mode `libssh2_sftp_shutdown` may return `LIBSSH2_ERROR_EAGAIN` when it calls `_libssh2_channel_free`. Since `ssh2` does not implement an async client, blocking mode is necessary so that `libssh2` retries until it is able to free the channel, otherwise memory associated with the channel will "leak" until `libssh2_session_free` is called when the session is dropped.

Unfortunately this means that the `ssh2` crate cannot really be used in an async client because dropping the `Sftp` can block indefinitely.

Resolving this probably requires a similar solution to https://github.com/alexcrichton/ssh2-rs/issues/220 - expose a method for async clients to use for shutting down the SFTP channel. In fact, it looks like we already have that method:
https://github.com/alexcrichton/ssh2-rs/blob/40d7d18415324205a9ba7ddd1c08b9f0fade42ec/src/sftp.rs#L425-L462

Is there any reason then for `Drop` to set the session to blocking before calling `libssh2_sftp_shutdown`? It seems like it would be preferable to "leak" the channel memory until free is called on the session instead of unexpectedly blocking an event loop if the user forgets to call the `shutdown` method.

@wez do you have any thoughts on this? it looks like you and @Ten0 were the last ones to touch this part of the code in these PRs:

1. https://github.com/alexcrichton/ssh2-rs/pull/156
2. https://github.com/alexcrichton/ssh2-rs/pull/196

Thanks!

**Comment by emesterhazy** [comment]
Actually, this slipped my mind, but I realize now that blocking here is the only safe thing to do since `libssh2` requires that the function is retried before any other functions are called if it returns `LIBSSH2_ERROR_EAGAIN`.

We probably need to do something similar for the `Drop` implementation for `Channel` so that async clients can call `libssh2_channel_free` and fall back to blocking during `Drop` to avoid corrupting the session state.

**Comment by Ten0** [comment]
> it looks like you and @Ten0 were the last ones to touch this part of the code

This commit fixed a use-after-free/double-free error, and I did not consider changing the behavior of changing to blocking to call free. Just made sure it was called only once. Also I do not use ssh2 in non-blocking mode.

> It looks like we already have that method

This crate seems to be somewhat inconsistent with regards to specifying whether it supports async directly:
Seems to specify we don't support directly:
 - https://github.com/alexcrichton/ssh2-rs/blob/40d7d18415324205a9ba7ddd1c08b9f0fade42ec/src/sftp.rs#L426-L427
 - https://github.com/alexcrichton/ssh2-rs/blob/124a367fff514875eca9e1b876b7476d31272984/src/lib.rs#L10-L12 (until https://github.com/alexcrichton/ssh2-rs/commit/7209aeba7fd7188ecb3727ac19416978ad99499d)

Seems to specify we do support it directly:
 - https://github.com/alexcrichton/ssh2-rs/blob/124a367fff514875eca9e1b876b7476d31272984/src/session.rs#L237-L245 (note the absence of `#[doc(hidden)]`)

It looks like the shutdown method should not be `#[doc(hidden)]` if it's intended that this crate is used directly in async contexts. On the other hand it looks like building an async abstraction over it before using it in async contexts would be a better design, so fixing another crate may be a better approach.

> It seems like it would be preferable to "leak" the channel memory until free is called on the session instead of unexpectedly blocking an event loop if the user forgets to call the shutdown method.

Hmm silently resource-leaking looks hard to diagnose as well. Panicking would also seem reasonable since we don't have async drops, and if 

> `libssh2` requires that the function is retried before any other functions are called if it returns `LIBSSH2_ERROR_EAGAIN`.

**Comment by emesterhazy** [comment]
Makes sense. I think it's fine if this crate does not support async

**Comment by emesterhazy** [comment]
> Seems to specify we do support it directly:
> 
> * https://github.com/alexcrichton/ssh2-rs/blob/124a367fff514875eca9e1b876b7476d31272984/src/session.rs#L237-L245
>    (note the absence of `#[doc(hidden)]`)

This is also a good point. I think until we make the updates above this method should be marked `#[doc(hidden)]` to avoid misleading users. As is, I don't think there's any way for users to use the structures in this crate in async code without risking memory leaks, session corruption, and unexpected blocking.

**Comment by harmic** [comment]
I should have mentioned this in #220 - I came across issues with Sftp as well. The `shutdown()` function isn't at all usable because you cannot call it back even if it does return EAGAIN. I should have raised a separate issue on that - sorry about that.

It looks to me like `shutdown()` might have been added specifically to allow async-ssh2 to expose shutdown without blocking (probably that is why it is hidden). 

The lack of any async drop capability in rust makes implementing any RAII pattern around this particularly painful in an async setting.

The solution I am currently using in my application is to put a wrapper around the Sftp objects, and then when they get dropped, I take the inner Sftp and send it to a dedicated dropping thread pool which operates synchronously. That way my async tasks do not have to wait around while these resources get released.

**Comment by crisidev** [comment]
I did exactly the same as @harmic using a wrapper around Sftp object. @emesterhazy would it be a good-enough solution also for you?

**Comment by harmic** [comment]
@crisidev you might want to take a look at #229. I am now not so sure my approach is working well.

**Comment by emesterhazy** [comment]
@crisidev Is there some code I could check out? I got busy with a few things but I'm starting to look at this again. I don't think we necessarily want to spawn additional threads to handle dropping the objects, at least not in this library. What I imagine is that we provide an interface that allows a higher level library (like tokio-libssh2, which I plan to write) to handle async Drop via a separate task like harmic is describing. That means we need to expose the ability to call the underlying libssh2 delete functions multiple times in order to retry them, and have some default "safe" fallback that allows objects to be cleaned up if the caller doesn't retry properly. We'd probably need to "poison" the session object in that case as well since once you start running an async `libssh2` command it seems that the only thing you can do is keep retrying it until it succeeds, otherwise you risk corrupting the session's state buffers.
